### PR TITLE
Edge fields extension and cluster undeployment improvement for edge

### PR DESCRIPTION
**Pull request by ankicabarisic**
- extending the edge to reflect node candidate properties 
- including fpga, gpu, cpuFrequency
- removing edge node source during the cluster removal
- promoting jobId change during cluster removal

### CODE REVIEW

**Comment by ankicabarisic** [comment]
The quality check is failing as the JSON field properties are moved to the constants to be reused and for easy maintainance. I guess we can ignore this quality check for this PR.

File: sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeDefinition.java [hunk 1]
```diff
@@ -13,38 +13,41 @@
 
 
 /**
- * Attributes defining a BYON node
+ * Attributes defining a EDGE node
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeDefinition.java]: Attributes defining **an** EDGE node

File: sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeDefinition.java [hunk 2]
```diff
@@ -13,38 +13,41 @@
...
  */
 @AllArgsConstructor
 @NoArgsConstructor
 @Getter
 @Setter
 @ToString(callSuper = true)
 public class EdgeDefinition {
-    @JsonProperty("name")
+    // Default constants
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeDefinition.java]: I think you can remove this comment

File: sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeNode.java [hunk 3]
```diff
@@ -29,54 +29,83 @@
...
+    public static final String JSON_ALLOCATED = "allocated";
+
+    // edge jobID corresponds to the ProActive job name
+    public static final String JSON_JOB_ID = "jobId";
+
+    public static final String JSON_SYSTEM_ARCH = "systemArch";
+
+    public static final String JSON_SCRIPT_URL = "scriptURL";
+
+    public static final String JSON_JAR_URL = "jarURL";
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeNode.java]: I think all these would be better in the EdgeDefinition class.
Also remove the comment `// Constants for JSON properties`

File: sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java [hunk 4]
```diff
@@ -134,6 +167,27 @@ public void setCores(Integer cores) {
...
+        return this;
+    }
+
+    /**
+     * Sets the CPU frequency in GHz.
+     * @param cpuFrequency CPU frequency in GHz
+     */
+    public void setCpuFrequency(Double cpuFrequency) {
+        this.cpuFrequency = cpuFrequency;
+    }
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java]: Replace by the @Setter annotation on the `cpuFrequency` field

File: sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java [hunk 5]
```diff
@@ -134,6 +167,27 @@ public void setCores(Integer cores) {
...
+
+    /**
+     * Gets the CPU frequency in GHz.
+     * @return cpuFrequency
+     */
+    public Double getCpuFrequency() {
+        return cpuFrequency;
+    }
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java]: Replace by the @Getter annotation on the `cpuFrequency` field

File: sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java [hunk 6]
```diff
@@ -155,6 +209,10 @@ public Integer getFpga() {
         return fpga;
     }
 
+    public void setFpga(Integer fpga) {
+        this.fpga = fpga;
+    }
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java]: Replace by the @Setter annotation on the `fpga` field

File: sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java [hunk 7]
```diff
@@ -171,6 +229,27 @@ public void setFpga(String machineType) {
...
+        this.gpu = gpu;
+    }
+
+    /**
+     * Gets the number of GPUs.
+     * @return gpu
+     */
+    public Integer getGpu() {
+        return gpu;
+    }
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java]: Same, replace by @Setter

File: sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java [hunk 8]
```diff
@@ -251,30 +330,33 @@ public boolean equals(Object o) {
         return Objects.equals(this.id, hardware.id) && Objects.equals(this.name, hardware.name) &&
                Objects.equals(this.providerId, hardware.providerId) && Objects.equals(this.cores, hardware.cores) &&
                Objects.equals(this.ram, hardware.ram) && Objects.equals(this.disk, hardware.disk) &&
-               Objects.equals(this.fpga, hardware.fpga) && Objects.equals(this.location, hardware.location) &&
-               Objects.equals(this.state, hardware.state) && Objects.equals(this.owner, hardware.owner);
+               Objects.equals(this.fpga, hardware.fpga) && Objects.equals(this.gpu, hardware.gpu) && // Added GPU comparison
+               Objects.equals(this.cpuFrequency, hardware.cpuFrequency) && // Added CPU frequency comparison
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/Hardware.java]: Remove comments.
Also consider using the @EqualsAndHashCode for all you equals and hascode implementations.
It is usually sufficient.

File: sal-common/src/main/java/org/ow2/proactive/sal/model/Image.java [hunk 9]
```diff
@@ -24,34 +24,49 @@
 @Entity
 @Table(name = "IMAGE")
 public class Image implements Serializable {
+    // Constants for JSON property names
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/Image.java]: Remove comment

File: sal-common/src/main/java/org/ow2/proactive/sal/model/LoginCredential.java [hunk 10]
```diff
@@ -23,13 +23,20 @@
 @NoArgsConstructor
 @Embeddable
 public class LoginCredential implements Serializable {
-    @JsonProperty("username")
+    // JSON property constants
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/LoginCredential.java]: Remove comment

File: sal-common/src/main/java/org/ow2/proactive/sal/model/NodeCandidate.java [hunk 11]
```diff
@@ -32,27 +32,49 @@
 @Entity
 @Table(name = "NODE_CANDIDATE")
 public class NodeCandidate implements Serializable {
+    // JSON Property Constants
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/NodeCandidate.java]: Remove comment

File: sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java [hunk 12]
```diff
@@ -73,40 +90,61 @@ public void setPrice(Double price) {
...
-    public Integer getNumberOfCores() {
-        return numberOfCores;
+    public Integer getCores() {
+        return cores;
     }
 
-    public void setNumberOfCores(Integer numberOfCores) {
-        this.numberOfCores = numberOfCores;
+    public void setCores(Integer numberOfCores) {
+        this.cores = numberOfCores;
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java]: Replace by @Setter and @Getter

File: sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java [hunk 13]
```diff
@@ -73,40 +90,61 @@ public void setPrice(Double price) {
...
+        this.cpuFrequency = cpuFrequency;
+    }
+
+    /**
+     * Gets the CPU frequency in GHz.
+     * @return cpuFrequency
+     */
+    public Double getCpuFrequency() {
+        return cpuFrequency;
+    }
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java]: Same

File: sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java [hunk 14]
```diff
@@ -73,40 +90,61 @@ public void setPrice(Double price) {
...
-    public Long getMemory() {
-        return memory;
+    public Long getRam() {
+        return ram;
     }
 
-    public void setMemory(Long memory) {
-        this.memory = memory;
+    public void setRam(Long memory) {
+        this.ram = memory;
```
Comment (AlexFalessi) [inline sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java]: Same. You can do that for every simple getter and setter.
I'll stop adding these kind of comments now.

**Review by AlexFalessi — APPROVED** [review]
Approved with minor non blocking comments

File: sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeNode.java (hunk 3 above)
Comment (ankicabarisic) [reply sal-common/src/main/java/org/ow2/proactive/sal/model/EdgeNode.java]: you are right :) done

File: sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java (hunk 14 above)
Comment (ankicabarisic) [reply sal-common/src/main/java/org/ow2/proactive/sal/model/NodeProperties.java]: thanks for this, I changed all model classes i was updating in this PR to use Lombok getters, setters and equal and hash. Probably same should be done for other classes in model but will let it for another time :)

**Pull request merged**